Automattic/wp-calypso
# CPT: Add trash button to post type list item

__Note:__ This pull request is in-progress. See "Open Questions" below.

This pull request seeks to enable a user to trash/delete a post from the custom post types listing screen.

__Open questions:__

Is this the best mechanism for displaying notices in response to asynchronous actions? A few alternatives to consider may be (a) hooking in to the delete success/failure actions from the notices reducer (how do we differentiate cases where we may not want to show the notice?) or (b) specifying a callback option to the `deletePost` action (callbacks are something of an anti-pattern in Redux action creators).

How do we flush or auto-update query results optimistically when deleting or trashing a post? Post queries are currently stored as an object mapping of serialized query to global post ID results. While the serialized query is simply a JSON string, converting it back to an object was never intended and (a) may be non-performant and (b) breaks the intention of the serializing being one-way.

Do we want to impose the "increment" action on the component triggering the action? We _could_ dispatch this automatically from the delete action, but only if using the `getState` second argument of the action thunk return value. We'd discussed this in the past with regards to editing posts and decided that this magic was undesirable.

__Testing instructions:__

Verify that deleting a non-trashed post or page causes the post to have its status changed to "trashed". Verify that deleting a trashed post or page or deleting any non-post/page causes the post to be permanently deleted. In both cases, verify that post counts in the filter bar are correspondingly changed, and that components are updated optimistically.

1. Navigate to [My Sites](http://calypso.localhost:3000/sites)
2. Select a site where a custom post type is enabled
3. Click the custom post type sidebar navigation item
4. Click Trash in the actions for a post
5. Note that...
 - The post disappears __(Not yet implemented)__
 - The filter bar post counts update
 - A success notice is shown after the delete completes
6. Navigate (manually) to change the post type slug in the path to either `post` or `page`
7. Note that your posts/pages are listed
8. Click Trash for any non-trashed post
9. Note that...
 - The post disappears __(Not yet implemented)__
 - The filter bar post counts update (decrementing the current filter status, incrementing "Trashed")
 - A success notice is shown after the delete completes, indicating that the post was sent to trash
